### ISSUE TITLE
ldc-1.7.0-beta1 (regression): Internal compiler error with implicit cast

### ISSUE THREAD

**Issue by rickrankin** [ldc-developers/ldc#2471]
I'm getting an internal compiler error with ldc-1.7.0-beta1 on code that works fine with dmd. When I compile the sample below, I get 

```
bugs\ldc-iterator.d(44): Error: Internal Compiler Error: Invalid struct cast from `Iterator!string` to `char`
```

but as I said, the code works as-is with dmd. If I changed line 44 to

```d
char ch = cast(immutable(char)) it++;
```

then the code compiles and works as expected with ldc, also. This is the first time I've tried to compile with ldc, but this code has been working with dmd for a long time. It seems like ldc is ignoring the opCast operator _unless_ the cast is explicit, but that's just a guess based on the error message and some ways I've tried to alter the definition of opCast to get ldc to accept it. So far, the only way I've been able to get it to work with ldc is by using an explicit cast.

```d
module example;

struct Iterator(T : T[])
{
  alias ThisType = Iterator!(T[]);
  alias ItemType = T;

  this(T[] container, size_t index = 0)
  {
    container_ = container;
    index_ = index;
  }

  U opCast(U)() const
    if (is(U == ItemType))
  {
    return this ? container_[index_] : ItemType.init;
  }

  ref ThisType opUnary(string op)()
    if (op == "++" || op == "--")
  {
    mixin(op ~ "index_;");
    return this;
  }

  @property ItemType value() const        { return container_[index_]; }

  alias value this;

  private T[] container_  = null;
  private size_t index_   = 0;
}

void main()
{
  import std.stdio : writefln;

  string test = "abcde";
  auto it = Iterator!string(test);

  writefln("ItemType = %s", typeid(Iterator!string.ItemType));
  writefln("it = %c", it);
  char ch = it++;
  writefln("ch = %c, it = %c", ch, it);
}
```

**Comment by kinke** [comment]
It has something to do with the postfix increment (and is not specific to Windows). `char ch = it; it++;` compiles, as does `char ch = ++it;` (the latter with different result obviously). I don't have my machine at my disposal, so I can't look into more details yet.

**Comment by JohanEngelen** [comment]
LDC 1.6.0 on godbolt works, but 1.7 doens't. So it's a 1.7 regression.

**Comment by kinke** [comment]
Most likely introduced by #2418; AST diff:

1.6:
```
DeclarationExp::toElem: (char ch = (Iterator!string __pitmp66 = it;) , it.opUnary() , __pitmp66.value();)
```

1.7:
```
DeclarationExp::toElem: (char ch = cast(char)((Iterator!string __pitmp2 = it;) , it.opUnary() , __pitmp2);)
```

Seems like the `opCast()` resolution doesn't work for a CommaExp.

**Comment by JohanEngelen** [comment]
Fixed by #2479
